robots/bebop.h

```diff
@@ -0,0 +1,1188 @@
+ */
+void
+Bebop::setPitch(const float pitch)
+{
+    assert(pitch >= -1.0f && pitch <= 1.0f);
+    if (m_IsConnected) {
+        DRONE_COMMAND(setPilotingPCMDPitch, round(pitch * 100.0f));
+        DRONE_COMMAND(setPilotingPCMDFlag, 1);
+    }
+}
+
+/*
+ * Set drone's roll, for banking left and right.
+ */
+void
+Bebop::setRoll(const float right)
+{
+    assert(right >= -1.0f && right <= 1.0f);
+    if (m_IsConnected) {
+        DRONE_COMMAND(setPilotingPCMDRoll, round(right * 100.0f));
+        DRONE_COMMAND(setPilotingPCMDFlag, 1);
+    }
+}
+
+/*
+ * Set drone's up/down motion for ascending/descending.
+ */
+void
+Bebop::setVerticalSpeed(const float up)
+{
+    assert(up >= -1.0f && up <= 1.0f);
+    if (m_IsConnected) {
+        DRONE_COMMAND(setPilotingPCMDGaz, round(up * 100.0f));
+    }
+}
+
+/*
+ * Set drone's yaw. The drone will turn really slowly.
+ */
+void
+Bebop::setYawSpeed(const float right)
+{
+    assert(right >= -1.0f && right <= 1.0f);
+    if (m_IsConnected) {
+        DRONE_COMMAND(setPilotingPCMDYaw, round(right * 100.0f));
+    }
+}
+
+/*
+ * Send the command to start video streaming.
+ */
+void
+Bebop::startStreaming()
+{
+    DRONE_COMMAND(sendMediaStreamingVideoEnable, 1);
+}
+
+/*
+ * Send the command to stop video streaming.
+ */
+void
+Bebop::stopStreaming()
+{
+    DRONE_COMMAND(sendMediaStreamingVideoEnable, 0);
+}
+
+/*
+ * Stops the drone from moving along all axes.
+ */
+void
+Bebop::stopMoving()
+{
+    if (m_IsConnected) {
+        setPitch(0);
+        setRoll(0);
+        setYawSpeed(0);
+        setVerticalSpeed(0);
+    }
+}
+
+/*
+ * Tells the drone to take a photo and store it.
+ */
+void
+Bebop::takePhoto()
+{
+    if (m_IsConnected) {
+        std::cout << "Taking photo" << std::endl;
+        DRONE_COMMAND(sendMediaRecordPicture, 0);
+    }
+}
+
+/*
+ * Adds an event handler for when the drone is taking off or landing, indicated
+ * by its parameter.
+ */
+void
+Bebop::setFlightEventHandler(const FlightEventHandler handler)
+{
+    m_FlightEventHandler = handler;
+}
+
+/*
+ * Waits for the drone to send a state-update command and returns the new
+ * state.
+ */
+inline eARCONTROLLER_DEVICE_STATE
+Bebop::getStateUpdate()
+{
+    m_Semaphore.wait();
+    return getState();
+}
+
+/*
+ * Returns the drone's connection state.
+ */
+inline eARCONTROLLER_DEVICE_STATE
+Bebop::getState()
+{
+    auto err = ARCONTROLLER_OK;
+    const auto state = ARCONTROLLER_Device_GetState(m_Device.get(), &err);
+    checkError(err);
+    return state;
+}
+
+/*
+ * Create the struct used by the ARSDK to interface with the drone.
+ */
+inline void
+Bebop::createControllerDevice()
+{
+    // create discovery device
+    static const auto deleter = [](ARDISCOVERY_Device_t *discover) {
+        ARDISCOVERY_Device_Delete(&discover);
+    };
+    auto derr = ARDISCOVERY_OK;
+    const std::unique_ptr<ARDISCOVERY_Device_t, decltype(deleter)> discover(ARDISCOVERY_Device_New(&derr), deleter);
```

Comment:
Woah that's cool! I had no idea unique_ptr had that functionality - super useful for wrapping C APIs like this

Reply:
Yeah it's quite cute :-). Also handy if you want to have a unique_ptr to an array on the heap.